bot/cogs/verification.py

```diff
@@ -32,26 +59,372 @@
+            return True
+
+        # Since `n_members` is a suspiciously large number, we will ask for confirmation
+        log.debug("Amount of users is too large, requesting staff confirmation")
+
+        core_devs = pydis.get_channel(constants.Channels.dev_core)
+        confirmation_msg = await core_devs.send(
+            f"<@&{constants.Roles.core_developers}> Verification determined that `{n_members}` members should "
+            f"be kicked as they haven't verified in `{KICKED_AFTER}` days. This is `{percentage:.2%}` of the "
+            f"guild's population. Proceed?",
+            allowed_mentions=MENTION_CORE_DEVS,
+        )
+
+        options = (constants.Emojis.incident_actioned, constants.Emojis.incident_unactioned)
+        for option in options:
+            await confirmation_msg.add_reaction(option)
+
+        def check(reaction: discord.Reaction, user: discord.User) -> bool:
+            """Check whether `reaction` is a valid reaction to `confirmation_msg`."""
+            return (
+                reaction.message.id == confirmation_msg.id  # Reacted to `confirmation_msg`
+                and str(reaction.emoji) in options  # With one of `options`
+                and not user.bot  # By a human
+            )
+
+        timeout = 60 * 5  # Seconds, i.e. 5 minutes
+        try:
+            choice, _ = await self.bot.wait_for("reaction_add", check=check, timeout=timeout)
+        except asyncio.TimeoutError:
+            log.debug("Staff prompt not answered, aborting operation")
+            return False
+        finally:
+            await confirmation_msg.clear_reactions()
+
+        result = str(choice) == constants.Emojis.incident_actioned
+        log.debug(f"Received answer: {choice}, result: {result}")
+
+        # Edit the prompt message to reflect the final choice
+        await confirmation_msg.edit(
+            content=f"Request to kick `{n_members}` members was {'authorized' if result else 'denied'}!"
+        )
+        return result
+
+    async def _kick_members(self, members: t.Set[discord.Member]) -> int:
+        """
+        Kick `members` from the PyDis guild.
+
+        Note that this is a potentially destructive operation. Returns the amount of successful
+        requests. Failed requests are logged at info level.
+        """
+        log.info(f"Kicking {len(members)} members from the guild (not verified after {KICKED_AFTER} days)")
+        n_kicked, bad_statuses = 0, set()
+
+        for member in members:
+            if is_verified(member):  # Member could have verified in the meantime
+                continue
+            with suppress(discord.Forbidden):
+                await member.send(KICKED_MESSAGE)  # Send message while user is still in guild
+            try:
+                await member.kick(reason=f"User has not verified in {KICKED_AFTER} days")
+            except discord.HTTPException as http_exc:
+                bad_statuses.add(http_exc.status)
```

Comment:
I'm not sure about the `gather`. The first time this runs, it will try to kick about 10k people. I honestly don't know what will happen, and whather `gather`ing the coros would make things any different. I was actually considering putting a short sleep after each request, too ~ but that wouldn't work with `gather`, I guess, since they'd sleep concurrently.

Reply:
We concluded a gather doesn't make sense due to rate limits. It probably isn't smart enough to know a separate request has triggered a rate limit, so it'll keep spamming the rest of the requests and rate limit the bot even harder. However, I haven't confirmed that. Regardless, we're going to be waiting a long time to finish all the requests so it doesn't make a difference if they are all scheduled at the start of an interval or spread out over the interval.